src/provider/provider_os_memory.c

```diff
@@ -0,0 +1,421 @@
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    if ((alignment > 0) && ((uintptr_t)addr & (alignment - 1))) {
+        if (os_config->traces) {
+            TLS_last_native_error = UMF_OS_RESULT_ERROR_ADDRESS_NOT_ALIGNED;
+            fprintf(stderr,
+                    "allocated address 0x%llx is not aligned to %zu (0x%zx) "
+                    "bytes\n",
+                    (unsigned long long)addr, alignment, alignment);
+        }
+        goto err_unmap;
+    }
+
+    ret = os_mbind(addr, size, os_config->numa_mode, os_config->nodemask,
+                   os_config->maxnode, os_config->numa_flags);
+    if (ret) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_BIND_FAILED;
+        if (os_config->traces) {
+            perror("binding memory to NUMA node failed");
+        }
+        goto err_unmap;
+    }
+
+    if (os_config->traces) {
+        int numa_node = -1;
+        int ret = os_get_mempolicy(&numa_node, NULL, 0, (void *)addr);
+        if (ret) {
+            fprintf(stderr, "cannot print assigned NUMA node (errno = %i)\n",
+                    errno);
+            perror("get_mempolicy()");
+        } else {
+            printf("alloc(%zu) = 0x%llx, assigned NUMA node = %i\n", size,
+                   (unsigned long long)addr, numa_node);
+        }
+    }
+
+    *resultPtr = addr;
+
+    return UMF_RESULT_SUCCESS;
+
+err_unmap:
+    (void)os_munmap(addr, size);
+    return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+}
+
+static enum umf_result_t os_free(void *provider, void *ptr, size_t size) {
+    if (provider == NULL || ptr == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_config_t *os_config = &os_provider->config;
+
+    int ret = os_munmap(ptr, size);
+    if (ret) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_FREE_FAILED;
+        if (os_config->traces) {
+            perror("memory deallocation failed");
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+void os_get_last_native_error(void *provider, const char **ppMessage,
+                              int32_t *pError) {
+    (void)provider; // unused
+
+    if (ppMessage == NULL || pError == NULL) {
+        assert(0);
+        return;
+    }
+
+    *pError = TLS_last_native_error;
+    *ppMessage = Native_error_str[*pError - UMF_OS_RESULT_SUCCESS];
+}
+
+enum umf_result_t os_get_recommended_page_size(void *provider, size_t size,
+                                               size_t *page_size) {
+    (void)size; // unused
+
+    if (provider == NULL || page_size == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_config_t *os_config = &os_provider->config;
+
+    if (page_size == NULL) {
```

Comment:
you've already checked the "page_size" param above

Reply:
Right, removed.